lib/internal/url.js

```diff
@@ -618,97 +644,183 @@ class URL {
   }
 
   set href(value) {
-    const valid = updateUrl(this.#context.href, updateActions.kHref, `${value}`, this.#onParseComplete);
-    if (!valid) { throw ERR_INVALID_URL(`${value}`); }
+    const href = bindingUrl.update(this.#context.href, updateActions.kHref, `${value}`);
+    if (!href) { throw ERR_INVALID_URL(`${value}`); }
+    this.#updateContext(href);
   }
 
   // readonly
   get origin() {
-    return this.#context.origin;
+    const protocol = StringPrototypeSlice(this.#context.href, 0, this.#context.protocol_end);
+
+    if (URLContext.SpecialProtocols.has(protocol)) {
+      if (protocol === 'file:') {
+        return 'null';
+      }
+      return `${protocol}//${this.host}`;
+    }
+
+    if (protocol === 'blob:') {
+      const path = this.pathname;
+      if (path.length > 0) {
+        try {
+          const out = new URL(path);
+          const blobProtocol = out.protocol;
+          if (URLContext.SpecialProtocols.has(blobProtocol)) {
+            return `${blobProtocol}//${out.host}`;
```

Comment:
Same comment here, you might be able to just expose the type and compare it with integer value `NOT_SPECIAL = 1`.

Reply:
You're right. I'm updating the PR. Thanks for the review @lemire